### PR TITLE
Add load_table array support for columnar method

### PR DESCRIPTION
**Pull request by xmnlab**
this PR:

- add load_table array support for columnar method
- fix tests for load_table columnar method
- include @jp-harvey load_table refactoring https://github.com/omnisci/pymapd/compare/master...jp-harvey:array-support?expand=1

### CODE REVIEW

File: pymapd/connection.py [hunk 1]
```diff
@@ -682,32 +682,33 @@ def load_table_columnar(
...
-
-            input_cols = _pandas_loaders.build_input_columnar(
-                data,
-                preserve_index=preserve_index,
-                chunk_size_bytes=chunk_size_bytes,
-                col_types=col_types,
-                col_names=col_names
-            )
-        else:
+        if not isinstance(data, pd.DataFrame):
```
Comment (xmnlab) [inline pymapd/connection.py:685]: I change logic to `not` to raise the exception ... and I move all the content one indentation back ( it doesn't need to be inside an if/else estructure.

File: tests/test_loaders.py [hunk 2]
```diff
@@ -14,6 +15,8 @@
 
 def assert_columnar_equal(result, expected):
 
+    assert len(result) == len(expected)
```
Comment (xmnlab) [inline tests/test_loaders.py:18]: depends on #309

**Comment by xmnlab** [comment]
it is ready for review. thanks!

**Comment by xmnlab** [comment]
thanks for the feedback @jp-harvey !
I will work on that. also I will improve the test to add the cases you commented. thanks!!

**Comment by xmnlab** [comment]
@jp-harvey I fixed some issues and add more tests.

I am still working to fix some issues.

just checking, for fixed length, `ColumnDetails` doesn't have this information, so probably it should be treat in follow up PR, is that right?

**Comment by xmnlab** [comment]
locally the tests for nullable cases worked (waiting for CI now).
probably would be good in a future do a refactoring in this code.

**Comment by xmnlab** [comment]
@jp-harvey it is ready again for review ... let me know if I can add any other case for the tests

**Comment by jp-harvey** [comment]
@xmnlab I can't actually add another review for some reason. Looks pretty good, everything has been sorted out from what I can see except empty arrays not being inserted when there's only one column:
```
data = [
    {'ary': [2,3,4]},
    {'ary': [4444]},
    {'ary': []},
    {'ary': None},
    {'ary': [2,3,4]},
    ]
```
results in:
```
ary
2,3,4
4444
null
2,3,4
```
The `[]` values are not being inserted as empty arrays, there should be five rows not four.

**Comment by xmnlab** [comment]
mm interesting .. because there is a test for empty array: https://github.com/omnisci/pymapd/pull/311/files#diff-c518af4aa4697fe7c2932799f1655f53R159

could you share the code you used?

**Comment by xmnlab** [comment]
@jp-harvey 

I tested it in a jupyter notebook .. and it worked: https://github.com/Quansight/ivan-sandbox/blob/master/pymapd/load-table-array-columnar.ipynb

if you could provide a complete example that is failing, I can work on that :)

**Comment by jp-harvey** [comment]
This is the DDL
```
CREATE TABLE arraytest1 (
  ary INTEGER[]);
```
And this is the code:
```
import pymapd
import pandas as pd
import os

username = os.environ.get('username', 'admin')
password = os.environ.get('password', 'HyperInteractive')

con = pymapd.connect(user=username, password=password, dbname='omnisci', host='localhost', port=6278, protocol='http')

tablename = 'arraytest1'

data = [
    {'ary': [2,3,4]},
    {'ary': [4444]},
    {'ary': []},
    {'ary': None},
    {'ary': [2,3,4]},
    ]

df = pd.DataFrame(data)
con.load_table_columnar(tablename, df)
```

**Comment by xmnlab** [comment]
@jp-harvey ... it worked here ... as you can see in the `test 4`

what could be the reason for this difference? 

currently I am using omniscidb version `'5.2.0dev-20200329-cecceef8da'`

and I am using connection protocol default (`binary`)

**Comment by xmnlab** [comment]
what is the version of omniscidb you are using?

**Comment by jp-harvey** [comment]
> what is the version of omniscidb you are using?

I'm using 5.2.1.

**Comment by xmnlab** [comment]
@jp-harvey I added some tests to test_integration that run `load_table` and check the result. it looks like it is working well :)

**Comment by jp-harvey** [comment]
Tested on our open source and cannot replicate previous issue, must be something specific to my local OmniSci install.

**Comment by xmnlab** [comment]
sounds good. I will open a new PR that. also I will create an issue for that. thanks @andrewseidl

**Comment by xmnlab** [comment]
@andrewseidl is it ready to be merged? let me know if I am missing anything for this PR :)

**Pull request merged**